tt_metal/impl/dispatch/dispatch_core_manager.hpp

```diff
@@ -0,0 +1,136 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "common/core_descriptor.hpp"
+
+namespace tt::tt_metal {
+
+// Dispatch core manager APIs track which cores are assigned to which dispatch functionality
+
+// A command queue is split into an issue queue and completion queue and needs two cores to interface with each queue.
+// One core dispatches commands to worker cores on the device
+
+// std::optional is used to determine whether core has been assigned
+// tt_cxy_pair is used over CoreCoord to denote location because remote device command queue interface cores are on the associated MMIO device
+struct dispatch_core_types_t {
+    std::optional<tt_cxy_pair> issue_queue_interface = std::nullopt;  // Pulls commands from the issue queue for a given command queue on a device
+    std::optional<tt_cxy_pair> completion_queue_interface = std::nullopt; // Pushes to completion queue for a given command queue on a device
+    std::optional<tt_cxy_pair> command_dispatcher = std::nullopt; // Relays work to worker cores on device that command is targeting. Currently for MMIO devices, command_dispatcher == completion_queue_interface
+    // TODO (abhullar): consider renaming these when supporting GH #3953 and #3954
+    std::optional<tt_cxy_pair> remote_processor = std::nullopt;    // Receives fast dispatch commands from ethernet router and relays to command_dispatcher
+    std::optional<tt_cxy_pair> remote_signaller = std::nullopt;    // Transmits data from command_dispatcher to ethernet router to send fast dispatch results off chip
+};
+
+inline CoreCoord get_physical_dispatch_coord(const tt_cxy_pair &logical_location) {
+    const metal_SocDescriptor &soc_desc = tt::Cluster::instance().get_soc_desc(logical_location.chip);
+    CoreCoord worker_core({
+            .x = static_cast<size_t>(soc_desc.worker_log_to_physical_routing_x.at(logical_location.x)),
+            .y = static_cast<size_t>(soc_desc.worker_log_to_physical_routing_y.at(logical_location.y)),
+    });
+    return worker_core;
+}
+
+class dispatch_core_manager {
+   public:
+    dispatch_core_manager &operator=(const dispatch_core_manager &) = delete;
+    dispatch_core_manager &operator=(dispatch_core_manager &&other) noexcept = delete;
+    dispatch_core_manager(const dispatch_core_manager &) = delete;
+    dispatch_core_manager(dispatch_core_manager &&other) noexcept = delete;
+
+    // Ugly to accept num HW CQs here but it is needed to pull the correct number of initially available dispatch cores for assignment
+    static dispatch_core_manager &get(uint8_t num_hw_cqs) {
+        static dispatch_core_manager inst = dispatch_core_manager(num_hw_cqs);
+        return inst;
+    }
+
+    const tt_cxy_pair &issue_queue_interface_core(chip_id_t device_id, uint16_t channel, uint8_t cq_id) {
```

Comment:
Add @brief for what the issue_queue_interface_core is.

Reply:
will do this for each api here